compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1188,6 +1192,74 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         }
     }
 
+  /** Returns a block assigning to `targets` the corresponding values in `rhs`.
+   *
+   *  @param targets A sequence of untyped trees on the LHS of a multiple assignment.
+   *  @param rhs The RHS of a multuple assignment.
+   */
+  private def typedPairwiseAssignments(
+      targets: List[untpd.Tree], rhs: Tree
+  )(using Context): Tree =
+    val s = mutable.ListBuffer[untpd.Tree]()
+    val u = formPairwiseAssignments(
+      s, targets, untpd.TypedSplice(rhs), rhs.tpe,
+      EmptyTermName)
+    u.map((b) => typed(b)).getOrElse(unitLiteral)
+
+  /** Appends to `statements` the assignments of `targets` to corresponding values in `rhs`.
+   *
+   *  @param statements A partially constructed sequence representing the desugaring of a
+   *  multiple assignment.
+   *  @param targets A sequence of expression representing variables to assign.
+   *  @param rhs The expression of the values to assign..
+   *  @param rhsType The type of `rhs`.
+   *  @param prefix A prefix for the name of syntactic vals created by the methods.
+   */
+  private def formPairwiseAssignments(
+      statements: mutable.ListBuffer[untpd.Tree],
+      targets: List[untpd.Tree],
+      rhs: untpd.Tree,
+      rhsType: Type,
+      prefix: TermName
+  )(using Context): Option[untpd.Block] =
+    val source = UniqueName.fresh(prefix)
+    val d = untpd.ValDef(source, untpd.TypeTree(rhsType), rhs)
+      .withSpan(rhs.span)
+      .withFlags(Synthetic)
+    statements.append(d)
+
+    rhsType.tupleElementTypes match {
+      case None =>
+        report.error(InvalidMultipleAssignmentSource(rhsType), rhs.srcPos)
+        None
+
+      case Some(e) if targets.length != e.length =>
+        report.error(MultipleAssignmentShapeMismatch(e.length, targets.length), rhs.srcPos)
+        None
+
+      case Some(sourceTypes) =>
+        /** The value to assign to the `i`-th source. */
+        def rhsElement(i: Int): untpd.Tree =
+          untpd.Select(untpd.Ident(source), nme.productAccessorName(i))
+            .withSpan(rhs.span)
+
+        /** Forms the assignments of the targets in the range [`i`, `targets.length`). */
+        @tailrec def loop(i: Int): Option[untpd.Block] =
+          if (i == targets.length) then
+            Some(untpd.Block(statements.toList, untpd.TypedSplice(unitLiteral)))
```

Comment:
I'm afraid I don't understand. Thrown out from where?

The block merely represents the desugaring of a sequence of assignments at one level of the tree. If this sequence happens to contain another multiple-assignment, it will also be represented by another block. For example:

```
(a, (b, c)) = (1, (2, c))
```

is meant to become

```
{ a = 1; { b = 2; c = 3; () }; () }
```

It is certainly possibly to optimize the desugaring to avoid the nested block, but I don't expect that to have different semantics. If I'm right, then is the optimization really worth?

Reply:
Nevermind, I understood what I got wrong!